drivers/sm_pwm_01c: initial import

fjmolinas opened the pull request:
### Contribution description

This PR adds the Amphenol sm-pwm-01c dust sensor driver. Although less complete and precise as PM sensors like PMS5003 or SPS30 it offers a low cost alternative to get some dust sensing input.

No bus interface is required just two gpio's are toggled as particles are detected with an infrared sensor.

Since the test uses the `progress_bar` module, the Kconfig configuration for that module was added.

### Testing procedure

- configure the progress bar: `make -C tests/progress_bar/` menuconfig

![image](https://user-images.githubusercontent.com/23060007/111979141-03378680-8b05-11eb-87aa-8f7b32a722ea.png)

- use Kconfig for module dependency resolution `TEST_CONFIG=1 make -C tests/progress_bar/ flash term`

```
2021-03-22 11:53:21,607 # main(): This is RIOT! (Version: 2021.04-devel-1047-g62d95-pr_driver_sm_pwm_01c)
2021-03-22 11:53:21,610 # Progress bar test application.
2021-03-22 11:53:21,611 # Progress bar 0 |██████████████████████████████████████████████████| 100%
2021-03-22 11:53:27,965 # Done!
2021-03-22 11:53:27,965 # Progress bar 1 |██████████████████████████████████████████████████|
2021-03-22 11:53:40,486 # Progress bar 2 |██████████████████████████████████████████████████|
2021-03-22 11:53:40,503 # Progress bar 3 |██████████████████████████████████████████████████|
2021-03-22 11:53:40,520 # Progress bar 4 |██████████████████████████████████████████████████|
2021-03-22 11:53:40,537 # Progress bar 5 |██████████████████████████████████████████████████|
2021-03-22 11:53:40,603 # Done!
```

- test the driver application `make -C tests/driver_sm_pwm_01c/ flash term`, test with and without Kconfig

```
# main(): This is RIOT! (Version: 2021.04-devel-1044-gfd36c-HEAD)
# sm_pwm_01c driver test application
# starting weighted average PM2.5 and PM10 measurements
#
# PM2.5 level: 1272 ug/m3|███████████              |
# PM10  level:  490 ug/m3|████                     |
````

leandrolanzieri commented on drivers/include/sm_pwm_01c.h:
```diff
@@ -0,0 +1,199 @@
+ * @internal
+ *
+ */
+typedef struct {
+    uint16_t buf[SM_PWM_01C_BUFFER_LEN]; /**< circular buffer memory */
+    size_t head;                         /**< current buffer head */
+} circ_buf_t;
```
Would it make sense to use http://doc.riot-os.org/cib_8h.html instead?

fjmolinas replied:
I didn't use it because the cib will note over-write the buffer content and here I do want to do so(I think since I made this choice a while ago :)).

leandrolanzieri commented on drivers/sm_pwm_01c/sm_pwm_01c.c:
```diff
@@ -0,0 +1,211 @@
+{
+    assert(dev);
+#ifdef MODULE_SM_PWM_01C_MA
+    unsigned int state = irq_disable();
```
The `irq_disable`/`irq_enable` could be placed just once outside the `ifdef` right?

leandrolanzieri commented on drivers/include/sm_pwm_01c.h:
```diff
@@ -0,0 +1,199 @@
+ * @param[out]  dev         Initialized device descriptor of SM_PWM_01C device
+ * @param[in]   params      The parameters for the SM_PWM_01C device
+ *
+ * @return                  0 on success
+ *                         -EIO GPIO error
```
```suggestion
 * @retval                  0 on success
 * @retval                  -EIO GPIO error
```

fjmolinas replied:
will do

leandrolanzieri commented on drivers/include/sm_pwm_01c.h:
```diff
@@ -0,0 +1,199 @@
+int sm_pwm_01c_init(sm_pwm_01c_t* dev, const sm_pwm_01c_params_t* params);
+
+/**
+ * @brief       Start continuos measurement of Large and Small particle
```
Here and bellow
```suggestion
 * @brief       Start continuous measurement of Large and Small particle
```

leandrolanzieri commented on drivers/sm_pwm_01c/Kconfig:
```diff
@@ -0,0 +1,59 @@
+# General Public License v2.1. See the file LICENSE in the top level
+# directory for more details.
+#
+config MODULE_SM_PWM_01C
+    bool "SM_PWM_01C Amphenol infrared dust sensor"
+    depends on HAS_PERIPH_GPIO
+    depends on HAS_PERIPH_GPIO_IRQ
+    depends on TEST_KCONFIG
+    select MODULE_CHECKSUM
+    select MODULE_PERIPH_GPIO
+    select MODULE_PERIPH_GPIO_IRQ
+    select MODULE_ZTIMER
+    select MODULE_ZTIMER_USEC
+    select MODULE_ZTIMER_PERIPH_TIMER
+    imply MODULE_SM_PWM_01C_MA
+
+config MODULE_SM_PWM_01C_MA
+    bool "Use a moving average for sensor values"
```
What about? 
```suggestion
menuconfig MODULE_SM_PWM_01C
    bool "SM_PWM_01C Amphenol infrared dust sensor"
    depends on HAS_PERIPH_GPIO
    depends on HAS_PERIPH_GPIO_IRQ
    depends on TEST_KCONFIG
    select MODULE_CHECKSUM
    select MODULE_PERIPH_GPIO
    select MODULE_PERIPH_GPIO_IRQ
    select MODULE_ZTIMER
    select MODULE_ZTIMER_USEC
    select MODULE_ZTIMER_PERIPH_TIMER

config MODULE_SM_PWM_01C_MA
    bool "Use a moving average for sensor values"
    depends on MODULE_SM_PWM_01C
    default y
```

fjmolinas replied:
qdone

leandrolanzieri commented on tests/driver_sm_pwm_01c/main.c:
```diff
@@ -0,0 +1,61 @@
+    sm_pwm_01c_t dev;
+
+    puts("sm_pwm_01c driver test application");
+    if (sm_pwm_01c_init(&dev,  &sm_pwm_01c_params[0]) != 0) {
```
```suggestion
    if (sm_pwm_01c_init(&dev, &sm_pwm_01c_params[0]) != 0) {
```

fjmolinas replied:
dome

fjmolinas commented:
@leandrolanzieri I addressed your change requests and some additional formatting

fjmolinas commented:
All green @leandrolanzieri!

leandrolanzieri commented on drivers/sm_pwm_01c/sm_pwm_01c.c:
```diff
@@ -0,0 +1,206 @@
+#include "sm_pwm_01c.h"
+#include "sm_pwm_01c_params.h"
+
+#define ENABLE_DEBUG (0)
```
```suggestion
#define ENABLE_DEBUG 0
```

leandrolanzieri reviewed (commented):
Some minor comments @fjmolinas, also can you check some warnings from the static tests that appear in the code view? Feel free to squash directly.

fjmolinas commented on drivers/include/sm_pwm_01c.h:
```diff
@@ -0,0 +1,208 @@
+ * @retval                    0 on success
+ * @retval                   -EIO GPIO error
+ */
+int sm_pwm_01c_init(sm_pwm_01c_t* dev, const sm_pwm_01c_params_t* params);
```
I thought I had uncrustigied the file.. seems not, did it now.

fjmolinas commented:
Should have checked the warnings before pushing...

fjmolinas commented:
Addressed comments and squashed, re-tested:

```
Welcome to pyterm!        PM2.5 level:  725 ug/m3|██████                   |
2021-04-07 11:34:08,973 # PM10  level:  362 ug/m3|███                      |
```

leandrolanzieri reviewed (approved):
ACK!

leandrolanzieri commented:
Oh sorry I just noticed @fjmolinas, can you fix the typos in the commit message of cb3c4082e1ad3498622b709cdcf1b3f346c54d52 ?

fjmolinas commented:
Done.

leandrolanzieri commented:
Unrelated CI issue. Re-triggering.

fjmolinas commented:
All green @leandrolanzieri!

fjmolinas commented:
Thanks for the review!
